src/service/history-impl.js

```diff
@@ -542,8 +542,10 @@ export class HistoryBindingNatural_ {
     return this.whenReady_(() => {
       const newState = this.mergeStateUpdate_(
           this.getState_(), opt_stateUpdate || {});
-      this.historyReplaceState_(newState, /* title */ undefined,
-          newState.fragment ? ('#' + newState.fragment) : undefined);
+      const url = (newState.url || '').replace(/#.*/, '');
```

Comment:
We also need to update the `replace_` for the viewer case. But I can take care of that. For now we need to protect the extra params to be send to the viewer.

For now please wrap the existing implt in https://github.com/ampproject/amphtml/blob/3d331aba53737f2464707dcff90ba644c107b01a/src/service/history-impl.js#L907 in a `if (!opt_stateUpdate.url)` and add a TODO for me for the else case. What's involved is checking the capability of the viewer and sending a different type of message.

Reply:
The viewer case is already sending the full opt_stateUpdate object in the message, is that an issue?

Reply:
Gotcha, done.